pkg/service/backup/worker_deduplicate.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (C) 2024 ScyllaDB
+
+package backup
+
+import (
+	"bytes"
+	"context"
+	"path"
+	"slices"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/scylladb/scylla-manager/v3/pkg/scyllaclient"
+	"github.com/scylladb/scylla-manager/v3/pkg/sstable"
+	"github.com/scylladb/scylla-manager/v3/pkg/util/parallel"
+)
+
+// Deduplicate handles the deduplicate stage of the backup process.
+// The implementation is expected to follow RFC document
+// https://docs.google.com/document/d/1EtGlF6UGNy34D_7QsnCheaukp3UwVObZU56PBdd0CQ8/edit#heading=h.jl2qbpcarwp9
+func (w *worker) Deduplicate(ctx context.Context, hosts []hostInfo, limits []DCLimit) (err error) {
+	f := func(h hostInfo) error {
+		w.Logger.Info(ctx, "Removing duplicated files from local snapshot", "host", h.IP)
+		err := w.deduplicateHost(ctx, h)
+		if err == nil {
+			w.Logger.Info(ctx, "Done deduplication", "host", h.IP)
+		}
+		return err
+	}
+
+	notify := func(h hostInfo, err error) {
+		w.Logger.Error(ctx, "Removing duplicated files failed on host", "host", h.IP, "error", err)
+	}
+
+	return inParallelWithLimits(hosts, limits, f, notify)
+}
+
+func (w *worker) deduplicateHost(ctx context.Context, h hostInfo) error {
+	if err := w.setRateLimit(ctx, h); err != nil {
+		return errors.Wrap(err, "set rate limit")
+	}
+
+	dirs := w.hostSnapshotDirs(h)
+	f := func(i int) (err error) {
+		d := dirs[i]
+		dataDst := h.Location.RemotePath(w.remoteSSTableDir(h, d))
+
+		var remoteFiles []string
+		listOpts := &scyllaclient.RcloneListDirOpts{
+			FilesOnly: true,
+			Recurse:   true,
+		}
+		if err := w.Client.RcloneListDirIter(ctx, h.IP, dataDst, listOpts, func(f *scyllaclient.RcloneListDirItem) {
+			remoteFiles = append(remoteFiles, f.Name)
+		}); err != nil {
+			return errors.Wrapf(err, "host %s: listing all files from %s", h.IP, dataDst)
+		}
+
+		// Iterate over all SSTable IDs and group files per ID.
+		ssTablesGroupByID := make(map[string][]string)
+		for _, file := range d.Progress.files {
+			id := sstable.ExtractID(file.Name)
+			ssTablesGroupByID[id] = append(ssTablesGroupByID[id], file.Name)
+		}
+
+		if err := w.basedOnUUIDGenerationAvailability(ctx, d, h, remoteFiles, ssTablesGroupByID); err != nil {
+			return errors.Wrap(err, "deduplication based on .crc32 content")
+		}
+
+		if err := w.basedOnCrc32Content(ctx, d, h, dataDst, ssTablesGroupByID); err != nil {
+			return errors.Wrap(err, "deduplication based on .crc32 content")
+		}
+
+		return nil
+	}
+
+	notify := func(i int, err error) {
+		d := dirs[i]
+		w.Logger.Error(ctx, "Failed to deduplicate host",
+			"host", d.Host,
+			"keyspace", d.Keyspace,
+			"table", d.Table,
+			"error", err,
+		)
+	}
+
+	return parallel.Run(len(dirs), 1, f, notify)
```

Comment:
Actually, we could change the limit to something higher... Chunks are partitioned by different dir already.
We set the rate limit on RClone as well.

I don't know what (and if) it would have impact on the deduplication process.  
Let's stay with this low value for now (it's actually the same parallelism as for uploadHost used in upload part), but it could be fine-tuned in the future if needed.

Reply:
I'm fine with that for now as #3928 is going to take care for that.